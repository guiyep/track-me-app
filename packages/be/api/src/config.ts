export const NAMES = {
  Lambda: 'MainApiLambda',
  ApiGateway: 'MainApiGateway',
};
