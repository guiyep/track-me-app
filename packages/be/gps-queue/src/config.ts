export const CONFIGURATIONS = {
  VisibilityTimeout: 300,
};
