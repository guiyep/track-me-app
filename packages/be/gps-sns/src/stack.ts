import { Construct } from 'constructs';
import type * as sqs from 'aws-cdk-lib/aws-sqs';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import { getConstants } from '@track-me-app/be-consts';
import { CfnOutput } from 'aws-cdk-lib';
const Consts = getConstants();

export class GpsSns extends Construct {
  constructor(
    scope: Construct,
    id: string,
    props?: {
      subscriptions?: sqs.IQueue[];
    },
  ) {
    super(scope, id);

    const topic = new sns.Topic(this, Consts.GpsSns.TOPIC_NAME);
    if (props?.subscriptions) {
      props.subscriptions.forEach((queue: sqs.IQueue) => {
        topic.addSubscription(new subscriptions.SqsSubscription(queue));
      });
    }

    new CfnOutput(this, `GpsSnsTopicArn`, {
      value: topic.topicArn,
    });
  }
}
