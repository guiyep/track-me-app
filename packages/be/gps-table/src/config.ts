export const NAMES = {
  DynamoGpsTable: 'GpsLocations',
};
