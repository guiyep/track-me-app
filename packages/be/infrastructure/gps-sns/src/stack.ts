import { Construct } from 'constructs';
import type * as sqs from 'aws-cdk-lib/aws-sqs';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import { getConstants } from '@track-me-app/be-consts';
import { CfnOutput } from 'aws-cdk-lib';
import type { AccessProps } from '@track-me-app/aws';
const Consts = getConstants();

// filterPolicy: {
//       eventType: sns.SubscriptionFilter.stringFilter({
//         allowlist: ['UserSignedUp'],
//       }),
//     },

type Messages = typeof Consts.GpsSns.MESSAGES;

export class GpsSns extends Construct {
  private readonly topic: sns.Topic;

  constructor(
    scope: Construct,
    id: string,
    props?: AccessProps & {
      queueSubscriptions?: {
        queue: sqs.IQueue;
        listenTo?: Messages[keyof Messages][];
      }[];
    },
  ) {
    super(scope, id);

    this.topic = new sns.Topic(this, Consts.GpsSns.TOPIC_NAME);

    props?.queueSubscriptions?.forEach(({ queue, listenTo }) => {
      this.topic.addSubscription(
        new subscriptions.SqsSubscription(queue, {
          filterPolicy: {
            eventType: sns.SubscriptionFilter.stringFilter({
              allowlist: listenTo,
            }),
          },
        }),
      );
      new CfnOutput(this, `GpsSnsTopicSubscribers`, {
        value: queue.queueArn + ' - ' + JSON.stringify(listenTo),
      });
    });

    props?.read?.forEach((grantable) => {
      this.topic.grantSubscribe(grantable);
    });

    props?.write?.forEach((grantable) => {
      this.topic.grantPublish(grantable);
    });

    props?.readAndWrite?.forEach((grantable) => {
      this.topic.grantSubscribe(grantable);
      this.topic.grantPublish(grantable);
    });

    new CfnOutput(this, `GpsSnsTopicArn`, {
      value: this.topic.topicArn,
      description: 'The ARN of the GPS SNS topic',
    });
  }
}
