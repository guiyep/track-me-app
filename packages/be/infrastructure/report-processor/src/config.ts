export const NAMES = {
  Lambda: 'ReportProcessorLambda',
};
