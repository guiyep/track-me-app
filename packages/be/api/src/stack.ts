import { Construct } from 'constructs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { LambdaRestApi } from 'aws-cdk-lib/aws-apigateway';
import { NAMES } from './config';
import * as path from 'path';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

export interface ApiProps {
  dependencies: {
    gpsSqs: string;
  };
}

export class Api extends Construct {
  public lambdaApi: NodejsFunction;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.lambdaApi = new NodejsFunction(this, NAMES.Lambda, {
      functionName: NAMES.Lambda,
      runtime: Runtime.NODEJS_LATEST,
      entry: path.join(__dirname, '../src/lambda/index.ts'),
      handler: 'handler',
    });

    new LambdaRestApi(this, NAMES.ApiGateway, {
      handler: this.lambdaApi,
      restApiName: NAMES.ApiGateway,
    });
  }
}
