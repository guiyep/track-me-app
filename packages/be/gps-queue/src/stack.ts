import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';
import { CONFIGURATIONS } from './config';
import { getConstants } from '@track-me-app/be-consts';
import type { AccessProps } from '@track-me-app/aws';
import { IFunction } from 'aws-cdk-lib/aws-lambda';

const Consts = getConstants();

export class GpsQueue extends Construct {
  public readonly queue: sqs.Queue;
  public readonly dlq: sqs.Queue;

  constructor(
    scope: Construct,
    id: string,
    props?: AccessProps & {
      listeners?: IFunction[];
    },
  ) {
    super(scope, id);

    this.dlq = new sqs.Queue(this, Consts.GpsLocationsQueue.DLQ, {
      visibilityTimeout: cdk.Duration.seconds(CONFIGURATIONS.VisibilityTimeout),
    });

    this.queue = new sqs.Queue(this, Consts.GpsLocationsQueue.QUEUE_NAME, {
      visibilityTimeout: cdk.Duration.seconds(CONFIGURATIONS.VisibilityTimeout),
      queueName: Consts.GpsLocationsQueue.QUEUE_NAME,
      deadLetterQueue: {
        queue: this.dlq,
        maxReceiveCount: 3,
      },
    });

    props?.read?.forEach((grantable) => {
      this.queue.grantConsumeMessages(grantable);
    });

    props?.write?.forEach((grantable) => {
      this.queue.grantSendMessages(grantable);
    });

    props?.readAndWrite?.forEach((grantable) => {
      this.queue.grantConsumeMessages(grantable);
      this.queue.grantSendMessages(grantable);
    });

    const lambdaEventSource = new lambdaEventSources.SqsEventSource(
      this.queue,
      {
        batchSize: 5,
      },
    );

    props?.listeners?.forEach((listener) => {
      this.queue.grantConsumeMessages(listener);
      this.queue.grantSendMessages(listener);
      listener.addEventSource(lambdaEventSource);
    });
  }
}
