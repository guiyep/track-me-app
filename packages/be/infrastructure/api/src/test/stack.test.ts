import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Api } from '../stack';
import { NAMES } from '../config';
import { omit } from '@track-me-app/testing';

describe('Api Stack', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');
  new Api(stack, 'TestApi');

  test('snapshot', () => {
    const template = Template.fromStack(stack);
    expect(omit(template.toJSON(), ['S3Key'])).toMatchSnapshot();
  });

  test('should create an API Gateway with correct properties', () => {
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::ApiGateway::RestApi', {
      Name: NAMES.ApiGateway,
    });
  });

  test('should create a Lambda function with correct properties', () => {
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Lambda::Function', {
      FunctionName: NAMES.Lambda,
      Runtime: 'nodejs18.x',
      Handler: 'index.handler',
    });
  });
});
