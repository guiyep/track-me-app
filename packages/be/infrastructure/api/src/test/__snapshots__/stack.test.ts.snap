// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api Stack snapshot 1`] = `
{
  "Outputs": {
    "TestApiMainApiGatewayEndpoint7A856E58": {
      "Value": {
        "Fn::Join": {
          "0": "",
          "1": {
            "0": "https://",
            "1": {
              "Ref": "TestApiMainApiGateway6BDC2E1D",
            },
            "2": ".execute-api.",
            "3": {
              "Ref": "AWS::Region",
            },
            "4": ".",
            "5": {
              "Ref": "AWS::URLSuffix",
            },
            "6": "/",
            "7": {
              "Ref": "TestApiMainApiGatewayDeploymentStageprod96AA5D61",
            },
            "8": "/",
          },
        },
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestApiMainApiGateway6BDC2E1D": {
      "Properties": {
        "Name": "MainApiGateway",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "TestApiMainApiGatewayANYADEB3614": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": {
              "0": "",
              "1": {
                "0": "arn:",
                "1": {
                  "Ref": "AWS::Partition",
                },
                "2": ":apigateway:",
                "3": {
                  "Ref": "AWS::Region",
                },
                "4": ":lambda:path/2015-03-31/functions/",
                "5": {
                  "Fn::GetAtt": {
                    "0": "TestApiMainApiLambdaE9AF4E93",
                    "1": "Arn",
                  },
                },
                "6": "/invocations",
              },
            },
          },
        },
        "ResourceId": {
          "Fn::GetAtt": {
            "0": "TestApiMainApiGateway6BDC2E1D",
            "1": "RootResourceId",
          },
        },
        "RestApiId": {
          "Ref": "TestApiMainApiGateway6BDC2E1D",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TestApiMainApiGatewayANYApiPermissionTestStackTestApiMainApiGatewayB643CDDAANY258DCA37": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": {
            "0": "TestApiMainApiLambdaE9AF4E93",
            "1": "Arn",
          },
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": {
            "0": "",
            "1": {
              "0": "arn:",
              "1": {
                "Ref": "AWS::Partition",
              },
              "10": "/*/",
              "2": ":execute-api:",
              "3": {
                "Ref": "AWS::Region",
              },
              "4": ":",
              "5": {
                "Ref": "AWS::AccountId",
              },
              "6": ":",
              "7": {
                "Ref": "TestApiMainApiGateway6BDC2E1D",
              },
              "8": "/",
              "9": {
                "Ref": "TestApiMainApiGatewayDeploymentStageprod96AA5D61",
              },
            },
          },
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestApiMainApiGatewayANYApiPermissionTestTestStackTestApiMainApiGatewayB643CDDAANYDAC50619": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": {
            "0": "TestApiMainApiLambdaE9AF4E93",
            "1": "Arn",
          },
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": {
            "0": "",
            "1": {
              "0": "arn:",
              "1": {
                "Ref": "AWS::Partition",
              },
              "2": ":execute-api:",
              "3": {
                "Ref": "AWS::Region",
              },
              "4": ":",
              "5": {
                "Ref": "AWS::AccountId",
              },
              "6": ":",
              "7": {
                "Ref": "TestApiMainApiGateway6BDC2E1D",
              },
              "8": "/test-invoke-stage/*/",
            },
          },
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestApiMainApiGatewayAccount33A9C9F6": {
      "DeletionPolicy": "Retain",
      "DependsOn": {
        "0": "TestApiMainApiGateway6BDC2E1D",
      },
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": {
            "0": "TestApiMainApiGatewayCloudWatchRole35536BA9",
            "1": "Arn",
          },
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "TestApiMainApiGatewayCloudWatchRole35536BA9": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": {
            "0": {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          },
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": {
          "0": {
            "Fn::Join": {
              "0": "",
              "1": {
                "0": "arn:",
                "1": {
                  "Ref": "AWS::Partition",
                },
                "2": ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              },
            },
          },
        },
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "TestApiMainApiGatewayDeploymentD7E89E0Fc3c7df44cff6467ee015e5c3d9edc64d": {
      "DependsOn": {
        "0": "TestApiMainApiGatewayproxyANY79151641",
        "1": "TestApiMainApiGatewayproxy58AE3F12",
        "2": "TestApiMainApiGatewayANYADEB3614",
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "TestApiMainApiGateway6BDC2E1D",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "TestApiMainApiGatewayDeploymentStageprod96AA5D61": {
      "DependsOn": {
        "0": "TestApiMainApiGatewayAccount33A9C9F6",
      },
      "Properties": {
        "DeploymentId": {
          "Ref": "TestApiMainApiGatewayDeploymentD7E89E0Fc3c7df44cff6467ee015e5c3d9edc64d",
        },
        "RestApiId": {
          "Ref": "TestApiMainApiGateway6BDC2E1D",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "TestApiMainApiGatewayproxy58AE3F12": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": {
            "0": "TestApiMainApiGateway6BDC2E1D",
            "1": "RootResourceId",
          },
        },
        "PathPart": "{proxy+}",
        "RestApiId": {
          "Ref": "TestApiMainApiGateway6BDC2E1D",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "TestApiMainApiGatewayproxyANY79151641": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": {
              "0": "",
              "1": {
                "0": "arn:",
                "1": {
                  "Ref": "AWS::Partition",
                },
                "2": ":apigateway:",
                "3": {
                  "Ref": "AWS::Region",
                },
                "4": ":lambda:path/2015-03-31/functions/",
                "5": {
                  "Fn::GetAtt": {
                    "0": "TestApiMainApiLambdaE9AF4E93",
                    "1": "Arn",
                  },
                },
                "6": "/invocations",
              },
            },
          },
        },
        "ResourceId": {
          "Ref": "TestApiMainApiGatewayproxy58AE3F12",
        },
        "RestApiId": {
          "Ref": "TestApiMainApiGateway6BDC2E1D",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TestApiMainApiGatewayproxyANYApiPermissionTestStackTestApiMainApiGatewayB643CDDAANYproxy2351E4AB": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": {
            "0": "TestApiMainApiLambdaE9AF4E93",
            "1": "Arn",
          },
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": {
            "0": "",
            "1": {
              "0": "arn:",
              "1": {
                "Ref": "AWS::Partition",
              },
              "10": "/*/*",
              "2": ":execute-api:",
              "3": {
                "Ref": "AWS::Region",
              },
              "4": ":",
              "5": {
                "Ref": "AWS::AccountId",
              },
              "6": ":",
              "7": {
                "Ref": "TestApiMainApiGateway6BDC2E1D",
              },
              "8": "/",
              "9": {
                "Ref": "TestApiMainApiGatewayDeploymentStageprod96AA5D61",
              },
            },
          },
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestApiMainApiGatewayproxyANYApiPermissionTestTestStackTestApiMainApiGatewayB643CDDAANYproxy935D4281": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": {
            "0": "TestApiMainApiLambdaE9AF4E93",
            "1": "Arn",
          },
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": {
            "0": "",
            "1": {
              "0": "arn:",
              "1": {
                "Ref": "AWS::Partition",
              },
              "2": ":execute-api:",
              "3": {
                "Ref": "AWS::Region",
              },
              "4": ":",
              "5": {
                "Ref": "AWS::AccountId",
              },
              "6": ":",
              "7": {
                "Ref": "TestApiMainApiGateway6BDC2E1D",
              },
              "8": "/test-invoke-stage/*/*",
            },
          },
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestApiMainApiLambdaE9AF4E93": {
      "DependsOn": {
        "0": "TestApiMainApiLambdaServiceRole612723F3",
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        },
        "FunctionName": "MainApiLambda",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": {
            "0": "TestApiMainApiLambdaServiceRole612723F3",
            "1": "Arn",
          },
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestApiMainApiLambdaServiceRole612723F3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": {
            "0": {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          },
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": {
          "0": {
            "Fn::Join": {
              "0": "",
              "1": {
                "0": "arn:",
                "1": {
                  "Ref": "AWS::Partition",
                },
                "2": ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              },
            },
          },
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": {
        "0": {
          "Assert": {
            "Fn::Not": {
              "0": {
                "Fn::Contains": {
                  "0": {
                    "0": "1",
                    "1": "2",
                    "2": "3",
                    "3": "4",
                    "4": "5",
                  },
                  "1": {
                    "Ref": "BootstrapVersion",
                  },
                },
              },
            },
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      },
    },
  },
}
`;
